File: src/No331.java
public class No331 {
    public static void main(String[] args) {
        // 利用图的性质 in == out && 在未遍历完之前out应该始终大于in
        String preorder = "1,2,#,#,3,#,#";
        System.out.println(preorder.length() == 1 && preorder.charAt(0) == '#');
        String[] chs = preorder.split(",");
        int len = chs.length;
        int in = 0;
        int out = 0;
        for(int i = 0; i < len; i ++){
            if(i == 0){
                out += 2;
                continue;
            }
            if(chs[i].equals("#")){
                in += 1;
            }else{
                in ++;
                out += 2;
            }
            if(i != len - 1 && in >= out) System.out.println(false);
        }
        System.out.println(in == out);
    }
}

File: src/No331.js
var isValidSerialization = function(preorder) {
    let stack = [];
    let chs = preorder.split(",");
    let len = chs.length;
    for(let i = 0; i < len; i ++){
        if(!isNaN(parseInt(chs[i]))){
            stack.push(parseInt(chs[i]));
        }else if(chs[i] === '#'){
            stack.push('#');
            while(stack.length>=3){
                let lenStack = stack.length;
                if(stack[lenStack-1] === '#' && stack[lenStack-2] === '#' && !isNaN(stack[lenStack-3])){
                    stack.pop();
                    stack.pop();
                    stack.pop();
                    stack.push('#');
                }else{
                    break;
                }
            }
        }
    }
    return stack.length == 1 && stack[0] == '#';
};

isValidSerialization("9,3,4,#,#,1,#,#,2,#,6,#,#");
